tests/factories/requests_.py

```diff
@@ -2,49 +2,49 @@
 from io import BytesIO
 
 import requests
-from factory import Factory, Faker
+from factory import Factory, Faker, post_generation
+
+# pylint:disable=no-self-argument,method-hidden
 
 
 class Response(Factory):
     class Meta:
         model = requests.Response
 
-    raw = None
-    json_data = None
     encoding = "utf-8"
     status_code = Faker(
         "random_element", elements=[200, 201, 301, 304, 401, 404, 500, 501],
     )
-    headers = None
-
-    @classmethod
-    def _adjust_kwargs(cls, **kwargs):
-        headers = kwargs["headers"] or {}
-        encoding = kwargs["encoding"]
-        raw = kwargs["raw"]
-        json_data = kwargs["json_data"]
 
-        if raw is None and json_data is not None:
-            raw = json.dumps(json_data)
+    @post_generation
+    def headers(response, _create, headers, **_kwargs):
+        """Lower-case all header names. Requests requires this."""
+        if headers:
+            response.headers = {key.lower(): value for key, value in headers.items()}
 
-            if not headers:
-                headers["Content-Type"] = f"application/json; charset={encoding}"
+    @post_generation
+    def json_data(response, _create, json_data, **_kwargs):
+        """Set raw and content-type if json_data is given."""
+        if json_data is not None:
+            response.headers.setdefault(
+                "content-type", f"application/json; charset={response.encoding}"
+            )
 
-        kwargs["raw"] = BytesIO(raw.encode(encoding)) if isinstance(raw, str) else raw
+            response.raw = json.dumps(json_data)
```

Comment:
This doesn't have any declared encoding, as the "raw" hook below takes effect after?

Reply:
Yes, the `@post_generation` methods get executed in the order they appear in the source. So this one is just turning `raw` into a unicode JSON string, and the next one encodes the string.